src/main/java/uk/gov/laa/ccms/caab/service/NotificationService.java

```diff
@@ -32,15 +46,91 @@ public Mono<NotificationSummary> getNotificationsSummary(String loginId, String
    *
-   * @param searchCriteria        The criteria on which to search.
-   * @param page                  The page number for pagination.
-   * @param size                  The size or number of records per page.
+   * @param searchCriteria The criteria on which to search.
+   * @param page           The page number for pagination.
+   * @param size           The size or number of records per page.
    * @return A Mono wrapping the notifications list based on search criteria.
    */
   public Mono<Notifications> getNotifications(NotificationSearchCriteria searchCriteria,
       final Integer page, final Integer size) {
     return soaApiClient.getNotifications(searchCriteria, page, size);
   }
 
+  /**
+   * Retrieve the content for a notification attachment. First attempt retrieval from
+   * S3, and if the document is not found attempt retrieval from EBS, then upload to S3 for future
+   * retrieval.
+   *
+   * @param notificationAttachmentId  The identifier of the notification attachment.
+   * @param loginId                   The login identifier for the user.
+   * @param userType                  Type of the user (e.g., admin, user).
+   * @return An Optional String containing the document content.
+   */
+  public Optional<String> getNotificationAttachment(String notificationAttachmentId,
+      String loginId, String userType) {
+
+    String content = null;
+    try {
+      content = getNotificationAttachmentFromS3(notificationAttachmentId);
+      if (!StringUtils.hasText(content)) {
+        log.warn("Notification attachment with ID '{}' retrieved from S3 has no content.",
+            notificationAttachmentId);
+      }
+    } catch (NoSuchKeyException ex) {
+      log.warn("Notification attachment with ID '{}' missing in S3. Attempting to retrieve "
+              + "from EBS instead.",
+          notificationAttachmentId);
+      content = getNotificationAttachementFromSoa(notificationAttachmentId, loginId, userType);
+      if (!StringUtils.hasText(content)) {
+        log.warn("Notification attachment with ID '{}' retrieved from EBS has no content.",
+            notificationAttachmentId);
+      }
+      uploadNotificationAttachmentToS3(notificationAttachmentId, content);
+    } catch (IOException ex) {
+      log.warn("Unable to process content for notification attachment with ID '{}'.",
+          notificationAttachmentId);
+    }
+
+    return Optional.ofNullable(content);
+  }
+
+  /**
+   * Retrieve the content of a notification attachment from S3.
+   *
+   * @param documentId The document identifier for the notification attachment.
+   * @return a String containing the content of the notification attachment.
+   * @throws IOException when the file content cannot be processed.
+   */
+  private String getNotificationAttachmentFromS3(String documentId) throws IOException {
+    return s3Template.download(bucketName, documentId).getContentAsString(StandardCharsets.UTF_8);
```

Comment:
Even though it's just a basic pass through, it might be better to create a uk.gov.laa.ccms.caab.client.S3Client to handle the actual call to S3. Good for re-use, and then the s3Template and AWS-specific exceptions can be encapsulated.

Reply:
Good point. Moved.